### PR TITLE
Set require_client_certificate in SSL context

### LINKED ISSUE #1002: Istio sidecar proxy is not correctly configured to enforce mTLS
My setup:
 - K8s cluster v1.6, running Calico for network
 - Istio-Auth based on latest `istio/istio`, tag at `a0268275a8704fe53fe4eec59b2adcdb7124492e`
 - BookInfo sample application.

I grab one of the backend pod IPs from k8s, and directly connect to it from outside the cluster.  Note that 192.168.18.139 is the `details` pod in my cluster.
```
$ curl --insecure https://192.168.18.139:9080/

    <html>
    <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <!-- Latest compiled and minified CSS -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap.min.css">

    <!-- Optional theme -->
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/css/bootstrap-theme.min.css">

    <!-- Latest compiled and minified JavaScript -->
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>

    <!-- Latest compiled and minified JavaScript -->
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.5/js/bootstrap.min.js"></script>

    </head>
    <title>Book details service</title>
    <body>
    <p><h2>Hello! This is the book details service. My content is</h2></p>
    <div>
<h4 class="text-center text-primary">Book Details</h4>
<dl>
<dt>Paperback:</dt>200 pages
<dt>Publisher:</dt> PublisherA
<dt>Language:</dt>English
<dt>ISBN-10:</dt>1234567890
<dt>ISBN-13:</dt>123-1234567980
</dl>
</div>
    </body>
    </html>
```

Expected behavior: The connection should be rejected, because it does not have a client cert and mTLS should be on.

### PR DESCRIPTION
**Pull request by spikecurtis**
Fixes #1002 

This fix causes pilot to configure Envoy to check client certificates on upstream proxy listeners.

### CODE REVIEW

File: proxy/envoy/route_test.go [hunk 1]
```diff
@@ -49,3 +49,11 @@ func TestSharedHost(t *testing.T) {
 		}
 	}
 }
+
+func TestBuildListenerSSLContext(t *testing.T) {
+	const dir = "/some/testing/dir"
+	context := buildListenerSSLContext(dir)
+	if !context.RequireClientCertificate {
+		t.Errorf("buildListenerSSLContext(%v) => Got RequireClientCertificate: true, expected false.", dir)
```
Comment (lookuptable) [inline proxy/envoy/route_test.go]: Should be `Got RequireClientCertificate: **false**, expected **true**.`

Comment (ldemailly) [inline proxy/envoy/route_test.go]: I thought go convention is to always put actual first

Comment (ldemailly) [inline proxy/envoy/route_test.go]: https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures

Comment (ldemailly) [inline proxy/envoy/route_test.go]: nvm - I should not read PR on phone, you are just asking to add ** ?

**Comment by kyessenov** [comment]
Please fix some linter errors (run `bin/check.sh` to check locally).

File: proxy/envoy/route_test.go (hunk 1 above)
Comment (lookuptable) [inline proxy/envoy/route_test.go]: No. I meant that the actual result is **false**.

Comment (spikecurtis) [inline proxy/envoy/route_test.go]: Fixed

**Comment by myidpt** [comment]
Thank you, Spike. It would be great if we can also test the case that the client doesn't present a cert and the server rejects the connection with the flag require_client_certificate set. (Or have we already covered it in this PR?)

**Comment by kyessenov** [comment]
@myidpt I filed issue #1005 and assigned it to you :)

**Comment by myidpt** [comment]
Oh cool.

**Pull request merged**